# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [348]:
# Imports section
import numpy as np
import pandas as pd
import sklearn as sk
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [349]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
dfsci = pd.read_csv("science_data_large.csv")
print(dfsci)

     Temperature °C  Mols KCL     Size nm^3
0               469       647  6.244743e+05
1               403       694  5.779610e+05
2               302       975  6.196847e+05
3               779       916  1.460449e+06
4               901        18  4.325726e+04
..              ...       ...           ...
995             894       847  1.545661e+06
996             327       982  6.737041e+05
997             791       213  3.477543e+05
998             769       553  8.684794e+05
999             919       452  8.476413e+05

[1000 rows x 3 columns]


## 2. Splitting the dataset

In [350]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X = dfsci.drop(columns=['Size nm^3'])  
y = dfsci[['Size nm^3']]           

# Split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

print(X_train)
#print(y_train)
print(X_test)
#print(y_test)

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]
     Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns]


## 3. Perform a Linear Regression

In [351]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()
model.fit(X_train, y_train)
score = model.score(X_test, y_test)

print(score)

coefficients = model.coef_
intercept = model.intercept_

print(f"coefficients: {coefficients}")
print(f"intercept: {intercept}")

#sample datapoint 
a = 100
b = 541
prediction = 866.14641*a +1032.69507*b - 409391.47958 
print(prediction)

0.8552472077276095
coefficients: [[ 866.14641337 1032.69506649]]
intercept: [-409391.47958341]
235911.19428999996


Write the linear equation of a slime: 
$$ y = 866.14641x_1 +1032.69507x_2 - 409391.47958 $$

Report on score and explain meaning: the score of 0.85 is representative of the amount of variance seen from the linear regression line. It's described as "the coefficient of determination of the prediction." The best possible score is 1 so 0.85 would be considered fairly good. For the input of 100 C and 541 Mols of KCL the predicted size was 235911

## 4. Use Cross Validation

In [352]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
shuffled = KFold(n_splits=5, random_state=42, shuffle=True)
scores = cross_val_score(model, X_train, y_train, cv = shuffled)
# Report on their finding and their significance
print(f"scores: {scores}")
print(f"standard deviation = {scores.std()}")
print(f"average = {scores.mean()}", )

scores: [0.86758302 0.84157293 0.84447006 0.85952305 0.88058427]
standard deviation = 0.014521183386787167
average = 0.8587466657377011


Write findings here: Even when the data is shuffled it still generally looks like a score of 0.84 to 0.88 so not a huge variance with the way kfold is moving things around. This tells us that the model is pretty stable and not under or over fitting too much. 

## 5. Using Polynomial Regression

In [353]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
# Report on the metrics and output the resultant equation as you did in Part 3.

polymodel = LinearRegression()
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X_train)
X_polytest = poly.transform(X_test)
polymodel.fit(X_poly, y_train)

scores_poly = cross_val_score(polymodel, X_poly, y_train, cv = shuffled)
score_poly = polymodel.score(X_polytest, y_test)
polycoefficients = polymodel.coef_
polyintercept = polymodel.intercept_

print(f"coefficients: {polycoefficients}")
print(f"intercept: {polyintercept}")
print(f"cross validation scores = {scores_poly}")
print(f"standard deviation = {scores_poly.std()}")
print(f"average = {scores_poly.mean()}", )

foo = np.array([[541, 100]])
foe = poly.transform(foo)
pred = polymodel.predict(foe)
print(f"prediction = {pred}")

coefficients: [[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
   2.00000000e+00  2.85714287e-02]]
intercept: [2.04771059e-05]
cross validation scores = [1. 1. 1. 1. 1.]
standard deviation = 0.0
average = 1.0
prediction = [[114977.71428741]]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Report on the score and interpret:$$y = 12.0000\,x_1 - 1.27197 \times 10^{-7}\,x_1^2 + 1.26194 \times 10^{-11}\,x_1x_2 + 2.00000\,x_2 + 2.85714 \times 10^{-2}\,x_2^2 + 2.04807 \times 10^{-5}$$

This model has a R^2 score of 1 indicating it is overfitting the data and likely not reliable for larger datasets. 

# Part 2. Chronic Kidney Disease Prediction via Classification

In [354]:
# Load the dataset. Then train and evaluate the classification models.

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd

# Load the dataset
ckd_data = pd.read_csv('ckd_feature_subset.csv')
X = ckd_data.drop(columns='Target_ckd')
y = ckd_data['Target_ckd']

# Initialize models
logistic_model = LogisticRegression(max_iter=1000)
svc_model = SVC()
knn_model = KNeighborsClassifier()

def evaluate_model(model, X, y, shuffled, name=None):
    name = model.__class__.__name__
    scores = cross_val_score(model, X, y, cv=shuffled, scoring='accuracy')
    
    print(f"Model: {name}")
    print(f"Cross-validation scores: {scores}")
    print(f"Average accuracy: {scores.mean():.5f}")
    print(f"Standard deviation: {scores.std():.5f}")
    print('-' * 80)

# Evaluate models
scores_logreg = evaluate_model(logistic_model, X, y, shuffled)
scores_svc = evaluate_model(svc_model, X, y, shuffled)
scores_knn = evaluate_model(knn_model, X, y, shuffled)


Model: LogisticRegression
Cross-validation scores: [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
Average accuracy: 0.85656
Standard deviation: 0.06627
--------------------------------------------------------------------------------
Model: SVC
Cross-validation scores: [0.90322581 1.         0.87096774 0.9        0.96666667]
Average accuracy: 0.92817
Standard deviation: 0.04760
--------------------------------------------------------------------------------
Model: KNeighborsClassifier
Cross-validation scores: [0.93548387 1.         0.87096774 0.86666667 0.96666667]
Average accuracy: 0.92796
Standard deviation: 0.05244
--------------------------------------------------------------------------------


In [355]:
from sklearn.neural_network import MLPClassifier

NN = MLPClassifier(max_iter=2000, solver='adam')
NNscores = cross_val_score(NN, X, y, cv=shuffled)

#learning_rate_init = 0.001
#hidden_layer_sizes = 100
#solver = 'adam'

print(NNscores)

[0.90322581 0.96774194 0.87096774 0.96666667 0.96666667]


only max iter = 2000: [0.90322581 0.96774194 0.87096774 0.96666667 0.96666667]\
init learn rate = 0.0001: [0.80645161 0.93548387 0.77419355 0.83333333 0.96666667]\
init learn rate = 0.0005: [0.93548387 1.         0.83870968 0.93333333 0.96666667]\
init learn rate = 0.01: [0.90322581 0.93548387 0.87096774 1.         0.96666667]\
init learn rate = 0.05: [0.90322581 1.         0.87096774 1.         0.96666667]\
solver = lbfgs: [1.         1.         0.93548387 1.         0.96666667]\
solver = sgd: [0.80645161 0.93548387 0.77419355 0.83333333 0.9       ]\
hidden layers = 25:[0.93548387 0.96774194 0.87096774 0.93333333 0.96666667]\
hidden layers = 50: [0.93548387 0.96774194 0.83870968 0.96666667 0.96666667]\
hidden layers: 200: [0.90322581 0.96774194 0.87096774 0.96666667 0.96666667\
hidden layers = 400: [0.90322581 1.         0.87096774 1.         0.96666667]\

## Results and Conclusion for Classification Experiments

Model: LogisticRegression\
Cross-validation scores: [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]\
Average accuracy: 0.85656\
Standard deviation: 0.06627\
--------------------------------------------------------------------------------\
Model: SVC\
Cross-validation scores: [0.90322581 1.         0.87096774 0.9        0.96666667]\
Average accuracy: 0.92817\
Standard deviation: 0.04760\
--------------------------------------------------------------------------------\
Model: KNeighborsClassifier\
Cross-validation scores: [0.93548387 1.         0.87096774 0.86666667 0.96666667]\
Average accuracy: 0.92796\
Standard deviation: 0.05244
--------------------------------------------------------------------------------

SVC and KNC(KNeighborsClassifier) were very close overall with logisticRegression being slightly behind. The best score was SVC only 0.001 avg score ahead of KNC.\
SVC also had a better standard deviation, so it's slightly more consistent. LogisticRegression is a bit behind, as its average scores were lower and it's standard deviation was also higher\
So we can concluse that the KNeightborsClassifier is the theoretically the best model here. 

## Results and Conclusion for Neural Network Experiments

| Configuration             | Score 1  | Score 2  | Score 3  | Score 4  | Score 5  | Mean Accuracy | Std Dev  |
|--------------------------|---------|---------|---------|---------|---------|---------------|----------|
| only max iter = 2000    | 0.90323 | 0.96774 | 0.87097 | 0.96667 | 0.96667 | 0.93505       | 0.04047  |
| init learn rate = 0.0001 | 0.80645 | 0.93548 | 0.77419 | 0.83333 | 0.96667 | 0.86323       | 0.07479  |
| init learn rate = 0.0005 | 0.93548 | 1.00000 | 0.83871 | 0.93333 | 0.96667 | 0.93484       | 0.05386  |
| init learn rate = 0.01   | 0.90323 | 0.93548 | 0.87097 | 1.00000 | 0.96667 | 0.93527       | 0.04547  |
| init learn rate = 0.05   | 0.90323 | 1.00000 | 0.87097 | 1.00000 | 0.96667 | 0.94817       | 0.05234  |
| solver = lbfgs          | 1.00000 | 1.00000 | 0.93548 | 1.00000 | 0.96667 | 0.98043       | 0.02592  |
| solver = sgd            | 0.80645 | 0.93548 | 0.77419 | 0.83333 | 0.90000 | 0.84989       | 0.05955  |
| hidden layers = 25      | 0.93548 | 0.96774 | 0.87097 | 0.93333 | 0.96667 | 0.93484       | 0.03515  |
| hidden layers = 50      | 0.93548 | 0.96774 | 0.83871 | 0.96667 | 0.96667 | 0.93505       | 0.04970  |
| hidden layers = 200     | 0.90323 | 0.96774 | 0.87097 | 0.96667 | 0.96667 | 0.93505       | 0.04047  |
| hidden layers = 400     | 0.90323 | 1.00000 | 0.87097 | 1.00000 | 0.96667 | 0.94817       | 0.05234  |

The highest mean accuracy was the solver = lbfgs at a 0.98 while the lowest mean accuracy was the solver = sgd at 0.84.\
The highest standard deviation was the init learn rate = 0.0001 with 0.074 and the lowest standard deviation was the solver = lbfgs with 0.02592. 

This tells us that solver = lbfgs should be the best performer with very high scores at very low standard deviation. While the extremely 
high scores could indicate overfitting, it also has very low standard deviation which is good, indicating it's a stable model. So it could
neccasary to add a validation set to help further evaluate it.

The worst performance seems to be solver = sgd as it has the lowest mean accuracy at 0.84 while also having a slightly higher than average 
standard deviation. The most inconsistent performance is the extremely low learning rate of 0.0001 which had a standard deviation of 0.074 in it's scores. 
This shows us that as you decrease the learning rate the the higher standard deviation of it's scores are, which seems like it would be counter intuitive, 
given that a more granular learn rate would seem like you'd get more precision rather than less. 

The last thing of note is that the effect of more hidden layers seems fairly low, as the average scores remained around 0.93 to 0.94. the standard deviation 
was also mostly similar with the exception of 25 hidden layers which is even lower than the others. This indicate that the number of hidden layers is a good place to save resources and you can stay at a somewhat lower number of hidden layers. 

It's also very possible that the method for my experiment is flawed as it accounts for each parameter in a vacuum. It could very well be the case that these parameters have very different effects depending on what other parameters they are combined with. It is also strange that in each different experimemnt similar scores come up such as 0.96667 or 0.90323 which might mean that the model is already extremly suited to the dataset by default and many parameters can be changed without major effects. 

 